libc/mman.c

```diff
@@ -0,0 +1,362 @@
+        // { MAP_UNINITIALIZED, SUPPORTED }
+    };
+
+    // Of the above flags, only MAP_FIXED is specified in POSIX.1-2001 and
+    // POSIX.1-2008.  However, most systems also support MAP_ANONYMOUS.
+
+    for (int i = 0; i < OE_COUNTOF(flags_table); ++i)
+    {
+        if (flags_copy & flags_table[i].flag)
+        {
+            if (flags_table[i].support == UNSUPPORTED)
+                OE_RAISE_MSG(
+                    OE_UNSUPPORTED,
+                    "unsupported `flag` value %d",
+                    flags_table[i].flag);
+
+            // Remove flag.
+            flags_copy &= ~flags_table[i].flag;
+        }
+    }
+
+    if (flags_copy)
+        OE_RAISE_MSG(OE_INVALID_PARAMETER, "invalid flag supplied");
+
+    if (length == 0)
+        OE_RAISE_MSG(OE_INVALID_PARAMETER, "length must be non zero");
+
+    // MAP_SHARED, MAP_SHARED_VALIDATE, MAP_PRIVATE are all treated as same
+    // since enclave is single process. Exactly one of them must be specified.
+    // They occupy the lower two bits of flags.
+    OE_STATIC_ASSERT(MAP_SHARED == 0x01);
+    OE_STATIC_ASSERT(MAP_PRIVATE == 0x02);
+    OE_STATIC_ASSERT(MAP_SHARED_VALIDATE == 0x03);
+
+    if (!(flags & 0x03))
+        OE_RAISE_MSG(
+            OE_INVALID_PARAMETER,
+            "`flags` must specify exactly one of MAP_SHARED or "
+            "MAP_SHARED_VALIDATE or MAP_PRIVATE");
+
+    if (flags & MAP_ANON || flags & MAP_ANONYMOUS)
+    {
+        // The fd argument is ignored; however, some implementations require fd
+        // to be -1 if MAP_ANONYMOUS (or MAP_ANON) is specified. The offset
+        // argument should be zero.
+        if (offset != 0)
+            OE_RAISE_MSG(
+                OE_INVALID_PARAMETER,
+                "offset` must be zero for anonymous mapping.");
+    }
+
+    result = OE_OK;
+done:
+    return result;
+}
+
+// See https://www.man7.org/linux/man-pages/man2/mmap.2.html for
+// semantics of mmap and munmap.
+void* oe_mmap(
+    void* addr,
+    size_t length,
+    int prot,
+    int flags,
+    int fd,
+    off_t offset)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    void* ptr = NULL;
+    void* vector = NULL;
+    oe_mapping_t* m = NULL;
+    size_t vector_length = 0;
+    int ret = 0;
+
+    OE_CHECK(_validate_mmap_parameters(addr, length, prot, flags, fd, offset));
+
+    _register_atexit_callback();
+
+    // length is rounded up to nearest page size.
+    OE_CHECK(oe_safe_round_up_u64(length, OE_PAGE_SIZE, &length));
+    OE_CHECK(oe_safe_round_up_u64(length / 8, 8, &vector_length));
+
+    // Allocate objects.
+    vector = (uint8_t*)calloc(vector_length, 1);
+    m = (oe_mapping_t*)malloc(sizeof(*m));
+    if ((ret = posix_memalign(&ptr, OE_PAGE_SIZE, length)) != 0)
+    {
+        // posix_memalign does not set errno (by spec).
+        // Set it ourselves.
+        oe_errno = ret;
+        OE_RAISE_MSG(
+            OE_OUT_OF_MEMORY, "posix_memalign failed with code %d", ret);
+    }
+
+    if (!vector || !ptr || !m)
```

Comment:
recommend moving this check up before line 191 so you don't call posix_memalign if out of memory

Reply:
Done.